### ISSUE TITLE
Minimum required set to get up and running on Heroku fails:  Package is not installed: ext-mongo-1.6.12.0  

### ISSUE THREAD

**Issue by holtkamp** [alcaeus/mongo-php-adapter#67]
As Heroku is quite fast with adopting PHP 7, I am looking at using this.

After reading https://github.com/alcaeus/mongo-php-adapter/issues/58 and https://github.com/alcaeus/mongo-php-adapter/issues/30 I came up with a short test but even the most basic ```composer.json``` fails to install:
```json
{
  "provide": {
    "ext-mongo": "1.6.12"
  },
  "require": {
    "alcaeus/mongo-php-adapter": "dev-master"
  }
}
```

Results in 

```
composer selfupdate

Updating to version 93501a5e3f8433e982da3518e13896ee4c0e816b.
    Downloading: 100%         
Use composer self-update --rollback to return to version d6d0435c5437111e42a123b06e4071e26ba7cb6e

composer install --ignore-platform-reqs
Loading composer repositories with package information
Installing dependencies (including require-dev)

                                                
  [InvalidArgumentException]                    
  Package is not installed: ext-mongo-1.6.12.0  
                                                

install [--prefer-source] [--prefer-dist] [--dry-run] [--dev] [--no-dev] [--no-plugins] [--no-custom-installers] [--no-autoloader] [--no-scripts] [--no-progress] [-v|vv|vvv|--verbose] [-o|--optimize-autoloader] [-a|--classmap-authoritative] [--ignore-platform-reqs] [--] [<packages>]...

```

Also changed ```replace``` with ```provide```, but same result...

Any idea where to look to at least get this installed?

**Comment by holtkamp** [comment]
Aah, nevermind, I now see https://github.com/alcaeus/mongo-php-adapter/blob/849de05fe57c4c1126036714c3b4dd9374edde05/composer.json#L20 already contains the provide part. 

Does this part of the documentations till apply?

```
If your project includes a library that requires ext-mongo you need to also specify a provide option in your composer.json: "provide": { "ext-mongo": "1.6.12" }

Due to a limitation in composer you need to specify this in the root package.```

**Comment by alcaeus** [comment]
Just a quick word about the `provide` directive in composer.json: that one is needed so you don't need to specify `--ignore-platform-reqs` during composer install. Not sure what exactly heroku needs to be satisfied all dependencies are present.

As for the error you are getting in that first example, I'll have to take a look at that.

**Comment by holtkamp** [comment]
Thanks for the response. Well, turns out my CLI was at php ```5.6.18``` when switching to ```7.0.3``` installation succeeds with:

```json
{
  "provide": {
    "ext-mongo": "1.6.12"
  },
  "require": {
    "alcaeus/mongo-php-adapter": "dev-master"
  }
}
```

But also without the ```provide``` part. What is the advised approach? Leave it at the ```mongo-php-adapter``` level?

**Comment by alcaeus** [comment]
The provide part is important when you add another package to the require. Go ahead and try adding the following packages to the `require` part:
```
"doctrine/mongodb-odm": "dev-master",
"doctrine/mongodb": "dev-master"
```

You'll find that you won't be able to install it with that `provide` directive but composer will complain about the missing `ext-mongo` dependency if you remove it. Of course, using `--ignore-platform-reqs` overrides this all so don't try it with that switch.

**Comment by holtkamp** [comment]
Mm, ok, will experiment a bit, maybe https://github.com/heroku/heroku-buildpack-php/issues/146 is not even an issue then...

**Comment by dzuelke** [comment]
@alcaeus should the directive not be `replace` instead of `provide` though?

**Comment by dzuelke** [comment]
Nevermind, just found #30 and #31. @alcaeus can you summarize your conversation with @seldaek and explain what the Composer limitations are that cause this? Is it because platform packages get special treatment internally?

**Comment by alcaeus** [comment]
Summary is: root package needs to specify `replace` but based on [my observations](https://github.com/alcaeus/mongo-php-adapter/issues/30#issuecomment-172926671) `provide` works while `replace` doesn't. The problem is due to all packages starting with "ext-" getting a special treatment and only being checked with `extension_loaded()` and `extension_version()`. I'll be taking a look at that once I've gotten Beta-1 out the door which should be in a day or two.

In an ideal world, `mongo-php-adapter` would specify a `replace` for `ext-mongo` and all projects wanting to use the adapter would just specify a `require` to `mongo-php-adapter` - no other stuff required.

**Comment by holtkamp** [comment]
Thanks for the detailed responses. The "ideal world" scenario sounds like a great place ;) Will run some additional tests to verify https://github.com/heroku/heroku-buildpack-php/issues/146 is an issue or not..

**Comment by dzuelke** [comment]
So I had a conversation with @seldaek yesterday and the tl;dr is that replace and provide should work pretty much the same way in this regard, and that both should be able to "override" PHP versions and extensions just like regular packages.

**Comment by alcaeus** [comment]
That sounds good - the question is, should it be possible to define this override somewhere other than the root package?

**Comment by dzuelke** [comment]
I believe the answer is yes. I have to create a few reproduce cases and then open a ticket showing which permutations work and which are broken.

**Comment by alcaeus** [comment]
@dzuelke Thanks for all the effort you're putting into this - I'll update the documentation to use `replace` instead of `provide` later.

**Comment by dzuelke** [comment]
What about this combination: mongo-php-adapter says "replace" for `ext-mongodb`, root project composer.json says "require" for `ext-mongodb`? Maybe that's the "correct" way, @alcaeus and @seldaek and @naderman? See composer/composer#2690

**Comment by jmikola** [comment]
If anything, this library would replace `ext-mongo` (i.e. the legacy driver). It depends on `ext-mongodb`, so I don't think you'd want to replace _that_.

**Comment by dzuelke** [comment]
Well yes of course @jmikola. Typo. M

**Comment by dzuelke** [comment]
Okay I can confirm that works, @alcaeus and @holtkamp (using the "composer-replace" branch since that still has it):

```
{
    "require": {
        "php": "^7.0",
        "alcaeus/mongo-php-adapter": "dev-composer-replace",
        "doctrine/mongodb": "dev-master"
    }
}
```

All is fine because `doctrine/mongodb` requires `ext-mongo`.

If no dependency requires that package, then you need to do it in the `composer.json` root:

```
{
    "require": {
        "php": "^7.0",
        "ext-mongo": "*",
        "alcaeus/mongo-php-adapter": "dev-composer-replace"
    }
}
```

This appears to be the preferred approach according to composer/composer#2690, and it works under all circumstances - both with `ext-foo`, as well as with "non-platform" packages (such as on Heroku, where "ext-foo" gets transformed to "heroku-sys/ext-foo").

**Comment by dzuelke** [comment]
PR here: #70

**Comment by alcaeus** [comment]
PR has been merged. @holtkamp Can you close this issue if it's confirmed fixed with the latest `dev-master` or follow up with any errors you encounter? Thanks!

**Comment by holtkamp** [comment]
mm, after removing my ```vendor``` dir, ```composer.lock``` and clearing composer caches, running ```composer install --ignore-platform-reqs``` with this ```composer.json```:

```json
{
  "require": {
    "php": "^7.0",
    "alcaeus/mongo-php-adapter": "dev-master",
    "doctrine/mongodb": "dev-master"
  }
}
```

succeeds. However when pushing to Heroku, the buildpack still complains:


```
git push heroku master         
Counting objects: 16, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (16/16), done.
Writing objects: 100% (16/16), 4.22 KiB | 0 bytes/s, done.
Total 16 (delta 10), reused 0 (delta 0)
remote: Compressing source files... done.
remote: Building source:
remote: 
remote: -----> Using set buildpack heroku/php
remote: -----> PHP app detected
remote: -----> Bootstrapping...
remote: -----> Installing system packages...
remote: 
remote:  !     ERROR: Failed to install system packages.
remote:        
remote:        Your platform requirements (for runtimes and extensions) could
remote:        not be resolved to an installable set of dependencies, or a
remote:        repository was unreachable.
remote:        
remote:        Full error information from installation attempt:
remote:        
remote:        > Loading repositories with available runtimes and extensions
remote:        > 
remote:        > Your requirements could not be resolved to an installable set of packages.
remote:        > 
remote:        >   Problem 1
remote:        >     - ext-mongo 1.6.12 requires php 5.5.* -> no matching package found.
remote:        >     - ext-mongo 1.6.12 requires php 5.6.* -> no matching package found.
remote:        >     - doctrine/mongodb dev-master requires ext-mongo ^1.5 -> satisfiable by ext-mongo[1.6.12].
remote:        >     - Installation request for doctrine/mongodb dev-master -> satisfiable by doctrine/mongodb[dev-master].
remote:        > 
remote:        
remote:        Please verify that all requirements for runtime versions in
remote:        'composer.lock' are compatible with the list below, and ensure
remote:        all required extensions are available for the desired runtimes.
remote:        
remote:        For reference, the following runtimes are currently available:
remote:        
remote:        PHP:  7.0.3, 7.0.2, 7.0.1, 7.0.0, 5.6.18, 5.6.17, 5.6.16, 
remote:        5.5.32, 5.5.31, 5.5.30
remote:        HHVM: 3.5.1
```


When doing the same attempt for the other described scenario, with a ```composer.json``` like:
```json
{
  "require": {
    "php": "^7.0",
    "alcaeus/mongo-php-adapter": "dev-master",
    "ext-mongo": "*"
  }
}
```

This results in:
```
git push heroku master                 
Counting objects: 20, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (20/20), done.
Writing objects: 100% (20/20), 4.59 KiB | 0 bytes/s, done.
Total 20 (delta 13), reused 0 (delta 0)
remote: Compressing source files... done.
remote: Building source:
remote: 
remote: -----> Using set buildpack heroku/php
remote: -----> PHP app detected
remote: -----> Bootstrapping...
remote: -----> Installing system packages...
remote: 
remote:  !     ERROR: Failed to install system packages.
remote:        
remote:        Your platform requirements (for runtimes and extensions) could
remote:        not be resolved to an installable set of dependencies, or a
remote:        repository was unreachable.
remote:        
remote:        Full error information from installation attempt:
remote:        
remote:        > Loading repositories with available runtimes and extensions
remote:        > 
remote:        > Your requirements could not be resolved to an installable set of packages.
remote:        > 
remote:        >   Problem 1
remote:        >     - ext-mongo 1.6.12 requires php 5.5.* -> no matching package found.
remote:        >     - ext-mongo 1.6.12 requires php 5.6.* -> no matching package found.
remote:        >     - Installation request for ext-mongo * -> satisfiable by ext-mongo[1.6.12].
remote:        > 
remote:        
remote:        Please verify that all requirements for runtime versions in
remote:        'composer.lock' are compatible with the list below, and ensure
remote:        all required extensions are available for the desired runtimes.
remote:        
remote:        For reference, the following runtimes are currently available:
remote:        
remote:        PHP:  7.0.3, 7.0.2, 7.0.1, 7.0.0, 5.6.18, 5.6.17, 5.6.16, 
remote:        5.5.32, 5.5.31, 5.5.30
remote:        HHVM: 3.5.1
remote:        
```

So, I am not sure whether this issue can be closed just yet, did I test correctly?

**Comment by alcaeus** [comment]
@holtkamp can you try `composer install` without `--ignore-platform-reqs`? The whole point of this `provide` vs. `replace` exercise is to get it working without using that nuclear option.

**Comment by dzuelke** [comment]
Yes @holtkamp the change is not merged yet for the buildpack. Please test according to the instructions at https://github.com/heroku/heroku-buildpack-php/issues/146#issuecomment-187451674

**Comment by holtkamp** [comment]
Mm, ok, sorry for my ignorance ;).

In both cases this results in problems when on PHP 7.0.3 CLI:

```
composer install
Loading composer repositories with package information
Installing dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Installation request for doctrine/mongodb dev-master -> satisfiable by doctrine/mongodb[dev-master].
    - doctrine/mongodb dev-master requires ext-mongo ^1.5 -> the requested PHP extension mongo is missing from your system.
```
And without ```doctrine/mongodb```
```
composer install
Loading composer repositories with package information
Installing dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - The requested PHP extension ext-mongo * is missing from your system. Install or enable PHP's mongo extension.
```

**Comment by dzuelke** [comment]
That's weird; if that happens on your box then it would have to happen on push to Heroku as well!?

**Comment by alcaeus** [comment]
Closing this - the composer issue is being tracked in composer/composer#5030 while the actual install issue is tracked in #30.
